csrc/preseg_passes/translate_no_reduction_matmul_to_mul_squeeze.cpp

```diff
@@ -0,0 +1,230 @@
+// sequence of MatmulOps with K=1.
+//
+// Step 2: Apply the translation.
+class NoReductionMatmulToMulSqueezeTranslator {
+ public:
+  NoReductionMatmulToMulSqueezeTranslator(Fusion* fusion) : fusion_(fusion) {}
+
+  void run() {
+    inspect();
+    translate();
+  }
+
+ private:
+  // Traverse through the fusion and gather all MatmulOps with K=1.
+  void inspect() {
+    const auto exprs = fusion_->exprs();
+
+    for (auto matmul : ir_utils::filterByType<MatmulOp>(exprs)) {
+      // Find the K dimension
+      auto in_a = matmul->inA();
+      auto in_b = matmul->inB();
+
+      auto domain_a = TensorDomain::noReductions(in_a->getLogicalDomain());
+      auto domain_b = TensorDomain::noReductions(in_b->getLogicalDomain());
+
+      auto ndims_a = domain_a.size();
+      auto ndims_b = domain_b.size();
+
+      if (ndims_a == 1 && ndims_b == 1) {
+        // Dot product. Should this be translated to a reduction as
+        // well?
+        continue;
+      }
+
+      IterDomain* k_id_a = domain_a.back();
+
+      if (!k_id_a->isBroadcast() || k_id_a->hasExpandedExtent()) {
+        continue;
+      }
+
+      // There should be no case where this extent is not one.
+      NVF_ERROR(
+          k_id_a->extent()->isOneInt(),
+          "Unexpected broadcast dimension: ",
+          k_id_a->toString());
+
+      no_reduction_matmul_.push_back(matmul);
+    }
+  }
+
+  // For each detected MatmulOp, apply the translation of mul+squeeze.
+  void translate() {
+    for (auto matmul : no_reduction_matmul_) {
+      auto in_a = matmul->inA();
+      auto in_b = matmul->inB();
+
+      // Given:
+      //
+      // out = MatmulOp(in_a, in_b)
+      //
+      // We broadcast in_a and in_b as necessary and do a pointwise
+      // multiplication and squeeze the K dimension. If there's no
+      // batch dimensions, either A or B would need to be
+      // broadcasted. Batch dimensions make things a little more
+      // complicated, but we align the batch dimensions of A and B by
+      // following the steps of torch.matmul
+      // (https://pytorch.org/docs/stable/generated/torch.matmul.html).
+
+      const auto ndims_a =
+          (int64_t)(TensorDomain::noReductions(in_a->getLogicalDomain())
+                        .size());
+      const auto ndims_b =
+          (int64_t)(TensorDomain::noReductions(in_b->getLogicalDomain())
+                        .size());
+      const auto batch_ndims_a = std::max(ndims_a - 2, (int64_t)0);
+      const auto matrix_ndims_a = std::min(ndims_a, (int64_t)2);
+      const auto batch_ndims_b = std::max(ndims_b - 2, (int64_t)0);
+      const auto matrix_ndims_b = std::min(ndims_b, (int64_t)2);
+
+      // Broadcast flags used to broadcast in_a. The number of false
+      // flags should be equal to batch_ndims_a + matrix_ndims_a. The
+      // vector is prepended with true if in_a misses some of the
+      // batch dimensions.
+      std::vector<bool> bc_flags_a;
```

Comment:
Can we use https://github.com/NVIDIA/Fuser/blob/aeb38d939a07f5b356379c5d8f2a09d8b2248622/csrc/ops/utils.cpp#L19 here?

Reply:
That can only be used for the batch dimensions. Not strictly required, but I wanted to insert just one broadcast op for both the batch and matrix dimensions, so that doesn't help here.